### PR TITLE
LPS-162476 Export miscellaneous modernized utils

### PR DESCRIPTION
**Pull request by kresimir-coko**

### CODE REVIEW

**Review by bryceosterhaus — APPROVED** [review]
LGTM! Just need to wait for CI to give the green light

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
It seems that 2 tests in [MapBase.js](https://github.com/liferay/liferay-portal/blob/master/modules/apps/map/map-common/test/liferay/MapBase.js) started failing after replacing `Liferay.Util.getGeolocation` with `getGeolocation`. I can't figure out how to fix it right now, but it's weird.

**Comment by bryceosterhaus** [comment]
@kresimir-coko did you check if it was mocked anywhere?

**Comment by kresimir-coko** [comment]
Yes I did check and replaced the mocks, @bryceosterhaus and replacing the mocks doesn't change the result. 
I'll get back onto checking it now, was preoccupied with the map fix thingy happening. 
Maybe I'll be smarter today.

[First mock](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/map/map-common/test/liferay/MapBase.js#L70-L74) and [second mock](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/map/map-common/test/liferay/MapBase.js#L130-L134).

I tried various combinations of replacing them, but nothing worked so far.

**Comment by kresimir-coko** [comment]
Hey @bryceosterhaus I decided to just remove the `getGeolocation` changes that were causing issues from this PR to get this one moving. I couldn't figure out the problem so I'll just tackle it separately on its own to not bog down this PR any further

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/__snapshots__/DocumentPreviewerWrapper.js.snap
Comment (bryceosterhaus) [inline modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/__snapshots__/DocumentPreviewerWrapper.js.snap:159]: Why did these change?

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (bryceosterhaus) [inline modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:22]: How come we are mocking `sub` here? I would think that we should expect sub to work as needed.

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/__snapshots__/DocumentPreviewerWrapper.js.snap
Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/__snapshots__/DocumentPreviewerWrapper.js.snap:159]: I'd call it an _improvement_.

@bryceosterhaus  If you compare the old mock to the new one, the old one was testing against a wrong implementation.

Just think about it, do we want `sub('document-x-of-x', [1, 1])` to display `document-1,1-of-x` or `document-1-of-1`?

The old test was simply written poorly.

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:22]: This new mock basically makes sure that all `x`s will be replaced with the arg, as that's how the test is written, so the mock supports that => `document-x-of-x` should be `document-1-of-1`, not `document-1,1-of-x`

Comment (bryceosterhaus) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:22]: That seems a little irrelevant to this test and the rest of this code though. Won't this just be another line of code to maintain without much purpose? Maybe I am missing something here, but what I'm trying to understand is how adding this mock improves our reliability on these tests, wouldn't we rather test against the real implementation of `sub`?

Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js]: You're right, the mock was super specific, so I replaced it (in https://github.com/liferay-frontend/liferay-portal/pull/2701/commits/d686d054f5e0d0856c1051829c77de44f2f0edba) with a mock that we already use across the codebase (but for `Liferay.Util.sub`).
The reason why there are some new mocks in some of these tests is because those tests relied on the globally available `Liferay.Util.utilName` mock defined in [liferay-frontend-projects](https://github.com/kresimir-coko/liferay-frontend-projects/blob/master/projects/npm-tools/packages/npm-scripts/src/jest/mocks/Liferay.js)

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (bryceosterhaus) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js]: If we totally remove the mocks to  `sub` that comes from frontend-js-web, what ends up being the issue? Is it just snapshots that need to be updated? I think ideally we wouldn't want to mock it at all because it is in a sense the "true" implementation of it. Whereas, for the global mock we had previously was used because we previously didn't have access to the sub function on the global namespace

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (bryceosterhaus) [inline modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:24]: Why do we need to mock the implementation at all for sub? If we don't mock it does it cause a break?

Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:24]: This test would pass without a mocked `sub`, but with updated snapshots.

- `document-1-of-1` would be `document-x-of-x`
- `the-document-Document-does-not-have-a-preview` would be `the-document-x-does-not-have-a-preview`
- etc.

Some tests would require way more adjustments if we were to altogether remove `sub` mocks.

The reason why all of a sudden we have some of these new mocks is because `Liferay.Util.sub` was mocked globally through our frontend-projects code. We could work on adding a sub mock there, I guess, so that it doesn't need to be defined in a lot of test files.

Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:24]: All our tests would have `x` instead of the expected value wherever we used `sub`. Another example is this test with a formatted date if you remember from the other day - [BasicInformation.js](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/analytics/analytics-reports-web/test/js/components/BasicInformation.js#L66), this file is not a part of this PR, but the next one. If we don't mock `sub`, the text that `getByText` is looking for is `published-on-x`, so this test would need to be changed. 
How do you imagine we handle these kinds of cases?

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (bryceosterhaus) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:24]: The reason I think the mock for `sub` is unhelpful here is because we are mocking a function that actually exists and works. Previously we had mocked `Liferay.Util.sub` because we didn't have a way to accurately test our implementation of that since we didn't have access to the global namespace of Liferay, so to handle that properly we added mocks for it. 

So for existing tests, it should be safe to update the snapshots from `document-1-of-1` to `document-x-of-x`. We don't really care that much whether it shows 1 or x because we are just making sure the snapshots are consistent and they reflect the actual implementation of `sub`.

For the case of [BasicInformation.js](https://github.com/liferay/liferay-portal/blob/master/modules/dxp/apps/analytics/analytics-reports-web/test/js/components/BasicInformation.js#L66), this is a case where we can add a single mock to this one test because it is dependent on the `x` value being properly replaced.

Another option would be to provide a global mock for `sub`. Currently `sub` is dependent on the transformation of language keys from `x-items` to `{0} items`. Since we don't have that in unit tests, we can add a global mock for `sub` to make things a bit more accurate. To add a global mock, I believe all you need to do is add a `__mocks__` folder and add a `sub.js` file at the same level of the current `sub.js` file

**Comment by kresimir-coko** [comment]
ci:stop

File: modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js
Comment (kresimir-coko) [reply modules/apps/digital-signature/digital-signature-web/test/js/components/document-previewer/DocumentPreviewerWrapper.js:24]: Alright, sounds good. I'll just remove any changes here relating to `sub` and leave them for a different PR then, and squash commits here.

I wasn't sure where this conversation was heading last week (sounded like you were referring to this specific test), so thanks for clearing it up.

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
@magjed4289 can you please check these failures please?

@bryceosterhaus it should look good now, can you give it a glance please?

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Review by bryceosterhaus — APPROVED** [review]
LGTM!

**Comment by kresimir-coko** [comment]
ci:reevaluate:1377401_5124

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:reevaluate:1337401_2246

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
Locally `dxp:apps:portal-search-tuning:portal-search-tuning-rankings-web:packageRunTest` passes successfully for me.

**Comment by kresimir-coko** [comment]
ci:test:acceptance-functional

**Review by bryceosterhaus — APPROVED** [review]
LGTM! Just need CI now

**Comment by mateusralv** [comment]
I'm taking a look [this](https://github.com/liferay-frontend/liferay-portal/pull/2701#issuecomment-1258165338) failures 🕵🏽‍♂️

**Comment by yanan** [comment]
I've checked the relevant test results [Comparation run](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=1990696328&testrayRunIdA=1990862061&statusB=2), @kresimir-coko the following tests look related, could you help take a look when you have time? Thanks in advance! cc @magjed4289 

1. Cannot open "Documents and Media" portlet > Options on Chrome 86, while it works on Chrome 100
- Control Panel > Sites > Add a new site using blank site template
- Go to Site > Site Builder > Pages > Add a new blank page
- Edit to add the "Documents and Media" portlet > click Options
- notice there is Options is not clickable on chrome86
chrome86+chrome100

2. Error about "Form Container". Here are the steps to reproduce attachment field problem on a form container:
- enable feature.flag.LPS-150277=true
- create a custom object
- create a new attachment field > choose "Upload or Select From Documents and Media Item Selector"
- publish the object
- Control Panel > Gogo Shell > Execute `scr:enable com.liferay.fragment.collection.contributor.inputs.InputsFragmentCollectionContributor`
- go to webpage > add "Form Container" fragment into the webpage
- Map the form in the Form container by selecting the object we just created
- notice error in browser console and "Select File" is not working
form-container-error

**Comment by kresimir-coko** [comment]
@yanan thank you, I will check them now.

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:reevaluate:1297401_4789

**Comment by yanan** [comment]
Tests failures are unrelated. Here is the [comparison run](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=1992355241&testrayRunIdA=1992762020&statusB=2) @kresimir-coko @magjed4289 

Flaky and run successful locally:
-LocalFile.WorkflowMetrics#ViewPerformanceByAssigneePage
- LocalFile.BlogsImage#CanUseBlogImageForDifferentEntries
- LocalFile.BlogsImage#CanViewCoverImageDetails

Existing tickets for the thers:
- LPS-147220
  - com.liferay.headless.admin.user.resource.v1_0.test.UserAccountResourceTest
- LPS-163282(LPS-163743) 
  - LocalFile.DatabasePartitioning#CanAddCompanyWithPortalInstancesAdminUI
  - LocalFile.DatabasePartitioning#CanDeleteCompany

**Pull request closed without merging**
